src/coreclr/src/vm/domainfile.cpp

```diff
@@ -1856,13 +1912,13 @@ BOOL DomainAssembly::CheckZapDependencyIdentities(PEImage *pNativeImage)
     CORCOMPILE_VERSION_INFO *pVersionInfo = pNativeImage->GetLoadedLayout()->GetNativeVersionInfo();
 
     // Check our own assembly first
-    GetAppDomain()->CheckForMismatchedNativeImages(&spec, &pVersionInfo->sourceAssembly.mvid);
+    GetLoaderAllocator()->CheckForMismatchedNativeImages(&spec, &pVersionInfo->sourceAssembly.mvid);
```

Comment:
I do not think that loader allocator is the right place to attach this to. If you have multiple (non-unlodable) AssemblyLoadContexts, all of them will share the one loader allocator, but it should be possible to load different versions of assemblies into each one.

In other words, loader allocator is lifetime unit; but this should be attached to assembly binding unit.

I know that the assembly loader is a mess and so the best way to do this may not be obvious.

Reply:
@davidwrighton could you help me find a home for this? I kicked it out of AppDomain based on [your feedback](https://github.com/dotnet/coreclr/pull/25281#discussion_r297394596), but it can't live in LoaderAllocator either.

Reply:
I have looked around:
- The unmanaged equivalent of `AssemblyLoadContext` is `ICLRPrivBinder` interface
- `ICLRPrivBinder` interface is implemented by several actual types
- There does not seem to be a good way to share stuff between these actual types

I think we need to:
- Create unmanaged `class AssemblyLoadContext`
- Put it between the managed `AssemblyLoadContext` and `ICLRPrivBinder` (e.g. the unmanaged `AssemblyLoadContext` is going to have `ICLRPrivBinder* m_pCLRBinder` field)
- Attach this structure to it

@elinor-fung @vitek-karas Thoughts?

Reply:
I think that would be reasonable. However, the assembly loader is such a tangle of indirection that I'm not sure how nice adding another layer would be. Potential related (existing) weirdness: the managed default `AssemblyLoadContext` class is created on demand, so default binder (and corresponding unmanaged `AssemblyLoadContext`) would exist without it, and the WinRT binder has no dedicated managed ALC representation

There is also the `ApplicationContext` instance (on each binder implementation). It has things like the cache of loaded managed assemblies, so I think it is kind of representative of the ALC as well. It might be fitting to have the mapping of the native image names/mvids there? It is currently only exposed on the concrete implementations (not `ICLRPrivBinder`) though.

The `CLRPrivBinderCoreCLR` and `CLRPrivBinderAssemblyLoadContext` implementations do have a lot of duplication (a bit less so for `CLRPrivBinderWinRT`). Adding a shared base for them might be nice in general.

Reply:
If I remember correctly we already discussed the fact that the interfaces (`IClrPrivBinder`) are not necessary and should be mostly removed (especially the fact that they are defined the "COM" way). I think that the solution here could be that we replace the `IClrPrivBinder` with an abstract base class, which can then store whatever additional state we need. It would also allow for an easier way to share code between the binders (as @elinor-fung notes, there's quite a bit of overlap in some cases).

Reply:
I have done some minimal refactoring to introduce AssemblyLoadContext base class in:  #32097